libs/core/langchain_core/messages/utils.py

```diff
@@ -242,3 +248,604 @@ def convert_to_messages(
+            "system", "human", "ai", ...) or as BaseMessage classes (e.g.
+            SystemMessage, HumanMessage, AIMessage, ...).
+        excl_types: Message types to exclude. Can be specified as string names (e.g.
+            "system", "human", "ai", ...) or as BaseMessage classes (e.g.
+            SystemMessage, HumanMessage, AIMessage, ...).
+        incl_ids: Message IDs to include.
+        excl_ids: Message IDs to exclude.
+
+    Returns:
+        A list of Messages that meets at least one of the incl_* conditions and none
+        of the excl_* conditions. If not incl_* conditions are specified then
+        anything that is not explicitly excluded will be included.
+
+    Raises:
+        ValueError if two incompatible arguments are provided.
+
+    Example:
+        .. code-block:: python
+
+            from langchain_core.messages import filter_messages, AIMessage, HumanMessage, SystemMessage
+
+            messages = [
+                SystemMessage("you're a good assistant."),
+                HumanMessage("what's your name", id="foo", name="example_user"),
+                AIMessage("steve-o", id="bar", name="example_assistant"),
+                HumanMessage("what's your favorite color", id="baz",),
+                AIMessage("silicon blue", id="blah",),
+            ]
+
+            filter_messages(
+                messages,
+                incl_names=("example_user", "example_assistant"),
+                incl_types=("system",),
+                excl_ids=("bar",),
+            )
+
+        .. code-block:: python
+
+            [
+                SystemMessage("you're a good assistant."),
+                HumanMessage("what's your name", id="foo", name="example_user"),
+            ]
+    """  # noqa: E501
+    messages = convert_to_messages(messages)
+    filtered: List[BaseMessage] = []
+    for msg in messages:
+        if excl_names and msg.name in excl_names:
+            continue
+        elif excl_types and _is_message_type(msg, excl_types):
+            continue
+        elif excl_ids and msg.id in excl_ids:
+            continue
+        else:
+            pass
+
+        # default to inclusion when no inclusion criteria given.
+        if not (incl_types or incl_ids or incl_names):
+            filtered.append(msg)
+            continue
+        if incl_names and msg.name in incl_names:
+            filtered.append(msg)
+            continue
+        elif incl_types and _is_message_type(msg, incl_types):
+            filtered.append(msg)
+            continue
+        elif incl_ids and msg.id in incl_ids:
+            filtered.append(msg)
+            continue
+        else:
+            pass
+
+    return filtered
+
+
+def merge_message_runs(
+    messages: Sequence[MessageLikeRepresentation],
+) -> List[BaseMessage]:
+    """Merge runs of Messages of the same type.
+
+    **NOTE**: ToolMessages are not merged, as each has a distinct tool call id that
```

Comment:
should we merge them if they have the same tool call ID?

Reply:
mm in my mind thats an invalid list of messages to begin with